### DISCUSSION TITLE
Adding Variations to the putlistingitem

### DISCUSSION THREAD

**Discussion by 200317314** [abuzuhri/Amazon-SP-API-CSharp#612] [Q&A] [answered]
I want to add a few color variations to my product that im listing on amazon, so far I have these 2 attributes ive added

`                if (!attrs.ContainsKey("variation_theme")) {
                        attrs.Add("variation_theme", new object[] {
                        new {
                            language_tag = "en_CA",
                            name = "COLOR",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }`

`

                if (!attrs.ContainsKey("parentage")) {
                    attrs.Add("parentage", new object[] {
                        new {
                            language_tag = "en_CA",
                            value = "parent",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }`

But how do I add the variant data to the ParameterPutListingItem() or ListingsItemPutRequest()

@TechyChap 
@abuzuhri 
@elonmusk
@jeffbezos

**Comment by TechyChap** [comment]
The process is to add first a parent product, then to create child products for each of your variations.

For the parent you'll need variation_theme and I've got parentage_level not parentage so you might want to check that.
For the child you'll need parentage_level "child", the variation_theme and child_parent_sku_relationship attribute with the link back to the parent_sku and child_relationship_type = variation. Obviously you'll also need to put the colour against the child.

It looks like I've also got child_parent_sku_relationship with a blank parent_sku against the parent but I'm not sure if that's strictly necessary.

**Reply by 200317314** [reply]
I am adding that parent and its telling me that "child_parent_sku_relationship" is missing, but when I look at the attributes I see it in there. This is what im using to add it.

`if (!attrs.ContainsKey("child_parent_sku_relationship")) {
                    attrs.Add("child_parent_sku_relationship", new object[] {
                        new {
                            language_tag = "en_CA",
                            parent_sku = "",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }`


![image](https://github.com/abuzuhri/Amazon-SP-API-CSharp/assets/31968510/1f0532b8-8a9e-4c9a-a40c-4d0c56f0317b)


Also how can I do multiple variation types? like Color and Size for the theme?

**Reply by TechyChap** [reply]
Not sure what's wrong with that. My advice would be to download one that is working (perhaps manually loaded into seller central) and compare that to your new one.

To do multiple variation types, just change the variation theme to "SIZE/COLOR" or one of the other enums mentioned in the schemas

**Reply by TechyChap** [reply, marked as answer]
Here the JSON for these attributes for a parent

```
          "parentage_level": [
            {
              "value": "parent",
              "marketplace_id": "A1F83G8C2ARO7P"
            }
          ],
          "child_parent_sku_relationship": [
            {
              "child_relationship_type": "variation",
              "marketplace_id": "A1F83G8C2ARO7P",
              "parent_sku": ""
            }
          ],
          "variation_theme": [
            {
              "name": "SIZE/COLOR"
            }
          ],
```

and for a child

```
 "parentage_level": [
            {
              "value": "child",
              "marketplace_id": "A1F83G8C2ARO7P"
            }
          ],
          "child_parent_sku_relationship": [
            {
              "child_relationship_type": "variation",
              "marketplace_id": "A1F83G8C2ARO7P",
              "parent_sku": "044701"
            }
          ],
          "variation_theme": [
            {
              "name": "SIZE/COLOR"
            }
          ],
```

**Reply by 200317314** [reply]
So I was able to get variations listed so its working now, my last problem is it does not like the enum "SIZE/COLOR" or really any enum that i checked for in the schemas.

variation_theme of just "COLOR" works great, not multi though

**Reply by TechyChap** [reply]
Once you set the variant type for a product you can't change it. So to use SIZE/COLOR you'd need to delete both the parent and children then add them back in. Note that you might need to wait a day after deletion for Amazon to purge it from their backend systems.

**Reply by 200317314** [reply]
This is for a a new parent, im not updating a child, the SIZE/COLOR theme is for a new parent that doesnt work, is there something else im missing?

**Reply by TechyChap** [reply]
For the parent, make sure you are not sending colour or size information.

**Reply by 200317314** [reply]
I ma getting this error
Message: Invalid enumerated value 'SIZE/COLOR' for attribute 'variation_theme'
Size and Color is not being sent in attributes

if (!attrs.ContainsKey("parentage_level")) {
                    attrs.Add("parentage_level", new object[] {
                        new {
                            language_tag = "en_CA",
                            value = "parent",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }

                if (!attrs.ContainsKey("child_parent_sku_relationship")) {
                    attrs.Add("child_parent_sku_relationship", new object[] {
                        new {
                            language_tag = "en_CA",
                            parent_sku = "",
                            value = "",
                            child_relationship_type = "variation",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }

                if (!attrs.ContainsKey("variation_theme")) {
                    attrs.Add("variation_theme", new object[] {
                        new {
                            language_tag = "en_CA",
                            name = (size.Equals("")) ? "COLOR" : "SIZE/COLOR",
                            marketplace_id = FikaAmazonAPI.Utils.MarketPlace.Canada.ID
                        }
                    });
                }

**Reply by 200317314** [reply]
issue is resolved its because PET_TOY does not have a SIZECOLOR enum

**Reply by wszwc** [reply]
What type of "attrs" is it? "attributes" can be directly equal to the json string I want to upload?
